presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -615,4 +640,52 @@ public MethodHandle resolveOperator(OperatorType operatorType, List<? extends Ty
         requireNonNull(functionRegistry, "functionRegistry is null");
         return functionRegistry.getScalarFunctionImplementation(functionRegistry.resolveOperator(operatorType, argumentTypes)).getMethodHandle();
     }
+
+    public static class TypeCompatibility
+    {
+        private final Optional<Type> commonSuperType;
+        private final boolean firstCanCoerceToSecond;
+
+        // Do not call constructor directly. Use factory methods.
+        private TypeCompatibility(Optional<Type> commonSuperType, boolean firstCanCoerceToSecond)
+        {
+            // Assert that: firstCanCoerceToSecond => commonSuperType.isPresent
+            // The factory API is designed such that this is guaranteed.
+            checkArgument(!firstCanCoerceToSecond || commonSuperType.isPresent());
+
+            this.commonSuperType = commonSuperType;
+            this.firstCanCoerceToSecond = firstCanCoerceToSecond;
+        }
+
+        private static TypeCompatibility explicit(Type commonSuperType, boolean firstCanCoerceToSecond)
```

Comment:
I think part of the problem with the names of these methods is that they are not symmetric. "Incompatible", which directly conveys the notion that types are not compatible with each other, is not the in the same category as "explicit/implicit", which describe the action or mode of constructing the TypeCompatibility object. To make them more coherent, you're need to go with something closer to that concept and name the methods in the vein of "compatibleButNotCoercible", "compatibleAndCoercible", etc. (names provided for illustration purposes, not that I like them)

Reply:
Potentially, I could remove `implicit` all together and have caller always specify to avoid this naming problem.